cmd/tools/gendynamicconfig/main.go

```diff
@@ -133,12 +132,48 @@ const Precedence{{.Name}} Precedence = {{.Index}}
 
 func generateType(w io.Writer, tp *settingType, prec *settingPrecedence) {
 	writeTemplatedCode(w, `
+{{ if .T.IsGeneric -}}
+type {{.P.Name}}TypedSetting[T any] setting[T, func({{.P.GoArgs}})]
+
+func New{{.P.Name}}TypedSetting[T any](key Key, convert func(any) (T, error), def T, description string) {{.P.Name}}TypedSetting[T] {
+	s := {{.P.Name}}TypedSetting[T]{
+		key:         key,
+		def:         def,
+		convert:     convert,
+		description: description,
+	}
```

Comment:
I could, but I want to allow different conversion functions. ConvertStructure is a convenient one but you might want to write your own for a specific use case. Or maybe more likely: expose a way to add hooks to ConvertStructure for specific field types

Reply:
Maybe have separate function that we pass the `converter`? Something like `New{{.P.Name}}.TypeSettingWithConverter`?

Reply:
True, it will be the most common (I converted a bunch of the maps and they all used it).

There's a question about defaults, though: should it default to mapstructure on top of the zero value or on top of the given default value? See this comment for the difference: https://github.com/temporalio/temporal/pull/5939/files#diff-7060a5c68f20a27374f24a6bfea103b9bcbed2e540c0afcd736a8c6a17c73304R286-R289

I think defaulting on top of the given default is usually what you want, but maybe not always? Also there's a restriction that it be shallow-copy-able without sharing structure (the specific value, not the type in general), which I don't think we can check at compile-time.

Reply:
Actually the shallow-copy-able check could be done at setting-creation-time since that should be static initialization time. It would be a bunch of custom reflection code though. Do you think it's worth having it check?

Reply:
I think it's ok as it is now.